tensorflow/core/common_runtime/threadpool_device.cc

```diff
@@ -43,7 +47,37 @@ ThreadPoolDevice::ThreadPoolDevice(const SessionOptions& options,
     : LocalDevice(options, Device::BuildDeviceAttributes(
                                name, DEVICE_CPU, memory_limit, locality)),
       allocator_(allocator),
-      scoped_allocator_mgr_(new ScopedAllocatorMgr(name)) {}
+      scoped_allocator_mgr_(new ScopedAllocatorMgr(name)) {
+#ifdef INTEL_MKL
+#ifdef _OPENMP
+  // Defaults for MKL optimized kernel execution.
+  if (getenv("KMP_BLOCKTIME") == NULL) {
+    // Explicitly set block time to reduce interference from spinning openmp
+    // threads
+    setenv("KMP_BLOCKTIME", "1", 1);
+  }
+
+  if (getenv("KMP_AFFINITY") == NULL && getenv("OMP_PROC_BIND") == NULL) {
+    // Bind threads to processors to reduce OS scheduling overhead
+    setenv("OMP_PROC_BIND", "true", 1);
+  }
+
+  const char* user_omp_threads = getenv("OMP_NUM_THREADS");
+  if (user_omp_threads == nullptr) {
+    // OMP_NUM_THREADS controls MKL's intra-op parallelization
+    // Default to available physical cores
+    const int mkl_intra_op = port::NumSchedulableCPUs();
+    const int ht = port::NumHyperthreadsPerCore();
+    omp_set_num_threads((mkl_intra_op + ht - 1) / ht);
```

Comment:
Can num_logical_cores not be divisible by num_hyper_threads? In other words, can each physical core have different number of hyper-threads?

Reply:
In hardware, no but the user can choose to explicitly disable some hyperthreads and not others using OS controls.